Intmdb-960-data federation database instance

govardhanpagidi opened the pull request:
Intmdb-960-data federation database instance
 new implementation, cfn tests are passed and tested with cfn template with private extension.


<!--
Check the boxes that apply. If you're unsure about any of them, don't hesitate to ask!
We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->
## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update

## Manual QA performed:

- [ ] cfn invoke for each of CRUDL/cfn test
- [ ] Updated resource in  [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples)
- [ ] Published to AWS private registry
- [ ] Used the template in [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples) to create and update a stack in AWS
- [ ] Deleted stack to ensure resources are deleted
- [ ] Created multiple resources in same stack
- [ ] Validated in Atlas UI
- [ ] Included screenshots

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code
- [ ] For CFN Resources: I have released by changes in the private registry and proved by change
  works in Atlas

## Further comments

govardhanpagidi commented:
<img width="944" alt="image" src="https://github.com/mongodb/mongodbatlas-cloudformation-resources/assets/15726683/4fe9ed8e-75c3-4463-9018-917b2f6a879d">

govardhanpagidi commented:
<img width="948" alt="image" src="https://github.com/mongodb/mongodbatlas-cloudformation-resources/assets/15726683/fa6d130d-60db-4847-a796-dae6cc45fd29">

Adelmar92 commented on cfn-resources/util/progressevent/failed_event.go:
```diff
@@ -23,6 +23,8 @@ import (
 func getHandlerErrorCode(response *http.Response) string {
 	switch response.StatusCode {
+	case http.StatusConflict:
+		return cloudformation.HandlerErrorCodeAlreadyExists
```
this is a complex case ghovardan, we avalueted mapping the conflict with the AlreadyExists before, we reialised that the StatusConflict is not always an Already exists, and each resource handles the AlreadyExists in different ways, some of them returns a 400 others a 200 with an empty body, so this might work for this resource, but not for others, this case we should handle en each resource

govardhanpagidi replied:
Ok, I ll move this to Resource level.

Adelmar92 commented on cfn-resources/federated-database-instance/mongodb-atlas-federateddatabaseinstance.json:
```diff
@@ -0,0 +1,355 @@
+    ],
+    "primaryIdentifier": [
+        "/properties/ProjectId",
+        "/properties/TenantName"
```
did you tested the read with a CFN stack?? i see that the profile is missing on the primaryIdentifier

govardhanpagidi replied:
Yes I tested with the example that is added, if we keep the Profile in PrimaryIdentifier we have to add return this as part of Get API which may not be required all the times to the user.

Adelmar92 replied:
i think you are confusing the primaryIdentifier with the readOnly, primaryIdentifiers can be createOnly and readOnly, if you set them to readOnly the user will be able to output them.
Also remember that only the primary identifiers reaches the read, that is why we added the profile, on your case is working because it arrives on nil and then is set as the default here..
<img width="662" alt="image" src="https://github.com/mongodb/mongodbatlas-cloudformation-resources/assets/69170650/882eab33-dca2-42b4-8aad-7a37159192a0">

if the user provides another profile, it will always use the default, remember that this is a behaviour that we had to fix on the end on the phase 1

govardhanpagidi replied:
I moved Profile to CreateOnly Properties and tested.

<img width="779" alt="image" src="https://github.com/mongodb/mongodbatlas-cloudformation-resources/assets/15726683/1677b0b7-d89c-4b90-9d06-ea07a9b57cf2">

Adelmar92 commented on cfn-resources/federated-database-instance/mongodb-atlas-federateddatabaseinstance.json:
```diff
@@ -0,0 +1,355 @@
+        "/properties/Profile"
+    ],
+    "required": [
+        "ProjectId"
```
remember to add all the required properties, on the CreateRequiredFields i see more, we need to add them here, because that affects the CDK

govardhanpagidi replied:
Sure

Adelmar92 commented on cfn-resources/federated-database-instance/cmd/resource/resource.go:
```diff
@@ -0,0 +1,492 @@
+// Create handles the Create event from the Cloudformation service.
+func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+	_, _ = logger.Debugf("Create Request: %+v\n", currentModel)
```
i think that we should only log errors, this logs just make errors harder to find

govardhanpagidi replied:
addressed

Adelmar92 commented on cfn-resources/federated-database-instance/cmd/resource/resource.go:
```diff
@@ -0,0 +1,492 @@
+		currentModel.Profile = aws.String(profile.DefaultProfile)
+	}
+
+	dataLakeTenantInput, _ := currentModel.setDataLakeTenant()
```
why are we skipping the error?

govardhanpagidi replied:
Removed Error, not required

Adelmar92 commented on cfn-resources/federated-database-instance/cmd/resource/resource.go:
```diff
@@ -0,0 +1,492 @@
+// Read handles the Read event from the Cloudformation service.
+func Read(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+	_, _ = logger.Debugf("Read Request: %+v", currentModel)
```
i would remove all of this logs

Adelmar92 commented on cfn-resources/federated-database-instance/cmd/resource/resource.go:
```diff
@@ -0,0 +1,492 @@
+}
+
+func handleError(response *http.Response, err error) (handler.ProgressEvent, error) {
+	return progress_events.GetFailedEventByResponse(fmt.Sprintf("Error during execution : %s", err.Error()),
```
is there a way to add in wich function (CREATE, DELETE, READ , LIST, UPDATE) was the error thrown?? is easy to follow on the testing stage

Adelmar92 commented on cfn-resources/federated-database-instance/cmd/resource/resource.go:
```diff
@@ -0,0 +1,492 @@
+
+		Storage: model.newDataFederationDataStorage(),
+	}
+	return dataLakeTenant, nil
```
the function defines an error but is returning nil every time

govardhanpagidi replied:
err not required so removed.

govardhanpagidi commented:
@Adelmar92 I've addressed review comments and tested.

<img width="779" alt="image" src="https://github.com/mongodb/mongodbatlas-cloudformation-resources/assets/15726683/a3622a8d-bf96-47a8-8c61-31413f5c03d5">

andreaangiolillo commented on examples/federated-database-instance/federated-database-instance.json:
```diff
@@ -0,0 +1,169 @@
+    "ProjectId": {
+      "Type": "String",
+      "Description": "Atlas Project Id",
+      "Default": ""
```
```suggestion
```

andreaangiolillo commented on examples/federated-database-instance/federated-database-instance.json:
```diff
@@ -0,0 +1,169 @@
+    "RoleID": {
+      "Type": "String",
+      "Description": "Unique identifier of the role that the federated database instance can use to access S3 Buckets. If necessary, use the Atlas API to retrieve the role ID. You must also specify the testS3Bucket",
+      "Default": "",
```
```suggestion
```

andreaangiolillo replied:
let's remove the empty defaults

govardhanpagidi replied:
Removed

andreaangiolillo commented on cfn-resources/util/util.go:
```diff
@@ -143,6 +149,75 @@ func NewMongoDBClient(req handler.Request, profileName *string) (*mongodbatlas.C
 }
 
+// NewAtlasClient func for creating atlas-go-sdk and mongodb-atlas-go client
+func NewAtlasClient(req *handler.Request, profileName *string) (*MongoDBClient, *handler.ProgressEvent) {
```
wasn't this added in another PR?

govardhanpagidi replied:
This is addressed by pulling latest from main

andreaangiolillo reviewed (commented):
we should add the contract test here

andreaangiolillo commented on .github/workflows/contract-testing.yaml:
```diff
@@ -24,6 +25,49 @@ jobs:
+          cat inputs/inputs_1_update.json
+          
+          make run-contract-testing
+          make delete-test-resources
```
you need to remove this job since we have the AWS policy issue right?

govardhanpagidi replied:
Removed

andreaangiolillo reviewed (commented):
test is failing

colm-quinn commented on cfn-resources/federated-database-instance/test/cfn-test-create-inputs.sh:
```diff
@@ -0,0 +1,143 @@
+
+
+echo -e "--------------------------------create key and key policy document starts ----------------------------\n"
+roleName="cfn-boto-df-role-${keyRegionUnderScore}"
```
I think this is the issue - can you try again using a role prefix of 'mongodb-atlas-' - permissions are restrict to a subset of roles for tests
